crates/sui-core/src/transaction_manager.rs

```diff
@@ -77,16 +87,26 @@ impl TransactionManager {
             if self.authority_store.effects_exists(&digest)? {
                 continue;
             }
+
+            let valid_effects = self.authority_store.get_valid_effects(&digest)?;
+
+            let effects = match &valid_effects {
+                Some(ValidEffectsInfo::Effects(tx_effects)) => Some(tx_effects.clone()),
+                _ => None,
+            };
+
             let missing = self
                 .authority_store
-                .get_missing_input_objects(&digest, &cert.data().data.input_objects()?)
+                .get_missing_input_objects(&digest, effects, &cert.data().data.input_objects()?)
                 .expect("Are shared object locks set prior to enqueueing certificates?");
 
             if missing.is_empty() {
-                self.certificate_ready(cert);
+                self.certificate_ready(cert, valid_effects);
                 continue;
             }
 
+            let mut entry = self.pending_certificates.entry(digest).or_default();
+            entry.valid_effects = valid_effects;
```

Comment:
I don't see logic reading this field. It seems we can drop this field and read cert & effect from db, or store certificate here as avoid reading from db when the transaction is ready to execute.